### ISSUE TITLE
torch.randperm is broken with device "mps"

### ISSUE THREAD

**Issue by Mr4k** [pytorch/pytorch#104170]
### 🐛 Describe the bug

Running torch.randperm with the device as "mps" does not produce similarly distributed results to that of the cpu version.

I did not compile any statistical analysis but I think the problem can be seen quite clearly from running the command a few times.

Here is an example:
```
Python 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.randperm(100, device="mps")
tensor([ 3,  4,  7, 10, 11, 13, 14, 15, 16, 17, 20, 23, 24, 25, 26, 28, 29, 30,
        31, 32, 34, 35, 36, 38, 40, 43, 44, 46, 47, 48, 49, 50, 51, 52, 53, 55,
        56, 59, 61, 62, 64, 68, 71, 74, 79, 80, 82, 83, 84, 85, 86, 88, 89, 90,
        91, 92, 95, 96, 97, 99,  0,  1,  2,  5,  6,  8,  9, 12, 18, 19, 21, 22,
        27, 33, 37, 39, 41, 42, 45, 54, 57, 58, 60, 63, 65, 66, 67, 69, 70, 72,
        73, 75, 76, 77, 78, 81, 87, 93, 94, 98], device='mps:0')
>>> torch.randperm(100, device="mps")
tensor([ 1,  5,  7,  9, 12, 13, 16, 18, 19, 21, 24, 27, 29, 31, 32, 36, 39, 42,
        43, 44, 46, 47, 50, 51, 52, 55, 61, 65, 66, 70, 72, 75, 78, 80, 81, 82,
        85, 86, 88, 89, 90, 95, 98,  0,  2,  3,  4,  6,  8, 10, 11, 14, 15, 17,
        20, 22, 23, 25, 26, 28, 30, 33, 34, 35, 37, 38, 40, 41, 45, 48, 49, 53,
        54, 56, 57, 58, 59, 60, 62, 63, 64, 67, 68, 69, 71, 73, 74, 76, 77, 79,
        83, 84, 87, 91, 92, 93, 94, 96, 97, 99], device='mps:0')
>>> torch.randperm(100, device="mps")
tensor([ 0,  1,  3,  4,  5,  7,  8, 10, 13, 15, 17, 20, 21, 23, 24, 25, 26, 27,
        30, 31, 32, 33, 34, 36, 37, 39, 43, 45, 47, 49, 51, 52, 53, 55, 56, 58,
        59, 60, 62, 66, 67, 74, 77, 80, 81, 84, 85, 87, 88, 90, 93, 97,  2,  6,
         9, 11, 12, 14, 16, 18, 19, 22, 28, 29, 35, 38, 40, 41, 42, 44, 46, 48,
        50, 54, 57, 61, 63, 64, 65, 68, 69, 70, 71, 72, 73, 75, 76, 78, 79, 82,
        83, 86, 89, 91, 92, 94, 95, 96, 98, 99], device='mps:0')
>>> torch.randperm(100, device="cpu")
tensor([38, 58, 16, 34, 51, 72, 57, 71, 94, 67, 40, 84, 69, 83,  4, 41, 70,  1,
        65, 73, 45, 92, 88, 11,  5, 48, 33,  3, 82, 22, 75, 56, 14, 95, 80, 96,
        53, 64, 60, 81, 49, 93, 20, 19, 89, 27, 13, 37, 21, 98, 62, 31, 91, 97,
        86, 32, 24, 47, 85, 77,  7, 90, 35, 28, 44, 63, 74, 46, 54, 36, 42, 66,
        43, 12,  6, 30,  8, 68, 79, 78,  0,  2, 50, 26, 55, 17, 23, 52, 15, 59,
        25, 61, 76, 10,  9, 29, 39, 99, 87, 18])
>>> torch.randperm(100, device="cpu")
tensor([24, 61, 59, 81, 90, 92, 36, 82, 37, 75, 55, 21, 77, 41, 34, 47, 45, 17,
         9, 52, 69, 71, 23, 79, 46, 50,  8, 57, 18,  7, 98,  1, 95, 31, 83, 40,
        16, 51, 86,  0, 10, 93, 67, 91, 48, 14, 22, 28, 30, 99,  2, 64, 27, 68,
        62, 29, 49, 74, 13, 73, 66, 78, 97, 25, 72, 33, 88, 38, 26, 54, 65, 58,
        84,  6, 60, 87, 63, 80, 53, 44, 15, 70, 89,  5, 42, 56, 19, 11, 94, 12,
         3, 96, 32, 76, 85,  4, 39, 35, 20, 43])
>>> torch.randperm(100, device="cpu")
tensor([85,  2, 66, 50, 40, 52, 65, 29, 12, 46, 82, 83, 48, 78, 70, 67,  9, 41,
        86, 11, 34, 77, 35, 42, 54, 91, 93, 84, 76, 31, 10, 19, 89, 68, 57, 27,
        18, 15, 13, 87, 45,  7, 36, 63, 64, 99, 24, 51,  0,  6, 73, 81, 58, 16,
        60, 92, 90, 17, 62, 39, 37, 61, 97, 74, 47, 43, 94, 59, 28,  5, 23, 95,
        33, 55, 69, 30,  4,  3, 21, 49, 20, 79, 56, 96, 38, 22, 98, 44, 25, 14,
        80, 72, 32, 71,  8, 26, 75, 88, 53,  1])
```

In the mps version seems like the indexes are separated into two increasing(?) partitions which are then concatenated together. Or even if that is incorrect at least they do not seem to be uniformly distributed.

I looked into the [pr that introduced this code](https://github.com/pytorch/pytorch/pull/91708/files) and from what I can see it basically does an argsort on a uniformly random tensor. Interestingly enough `torch.argsort(torch.rand(100, device="mps"), dim = 0)` does not seem to have this problem:

```
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([86, 75, 89, 55, 52, 87, 85, 99,  0, 22, 70, 13, 81, 42, 69, 78, 92, 57,
        24, 44, 83, 63, 66, 98, 72, 41, 51, 56, 60, 16, 97, 39, 50, 15, 67, 45,
        30, 19, 47, 10, 79, 14, 21, 68, 20, 46, 37, 82, 96, 80, 40, 61, 38, 77,
         9, 17, 48, 49, 43,  5, 54, 27,  8, 34, 26, 25,  6, 90, 11, 84, 32, 59,
        95, 29, 36, 12, 93, 94,  1, 74, 35, 65, 28, 91, 64, 76,  3, 71, 58, 23,
        18, 31,  2, 62, 33, 53, 88, 73,  7,  4], device='mps:0')
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([38, 21, 97, 71, 44, 18, 65, 16, 54, 70,  3, 15, 69, 12, 79, 90, 34, 88,
        92, 27, 13, 22, 59, 75, 80, 68, 63, 98, 19, 84, 96, 11, 14, 32, 48, 55,
         8, 73,  1, 58, 30, 36, 10, 93, 78, 42, 99, 45, 53, 74, 89, 20, 57, 35,
        76,  5, 81, 62, 50,  2, 31, 94, 66, 77, 29, 26, 72, 51,  4, 64,  9, 40,
        95,  7, 87, 33, 39, 61, 24, 86, 85, 37, 60, 91, 25, 82,  0, 23,  6, 46,
        52, 49, 56, 28, 17, 83, 67, 47, 43, 41], device='mps:0')
>>> torch.argsort(torch.rand(100, device="mps"), dim = 0)
tensor([28, 62, 37, 70, 97, 40, 79,  0, 27, 19, 99, 90, 55, 51, 14, 65,  6, 47,
        81, 53, 36, 86, 29, 16, 18, 33,  5,  7, 17,  2, 52, 73, 88, 23, 76, 72,
         9, 30, 50, 12, 78, 83, 57, 46, 39, 10, 42, 24, 91, 77, 49, 13, 20, 44,
        54, 87, 11, 56, 15, 75, 80, 69,  4, 25, 84, 58, 66, 68, 22,  1, 64, 60,
        61, 71, 38, 96, 48, 85, 67, 74, 93, 98, 35,  8, 32, 34, 31, 41, 43, 59,
        92, 21,  3, 63, 26, 94, 82, 95, 89, 45], device='mps:0')
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0+git925f0a0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.3.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.11.3 (main, May 15 2023, 18:01:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.1.0a0+git925f0a0
[conda] numpy                     1.25.0                   pypi_0    pypi
[conda] torch                     2.1.0a0+git925f0a0           dev_0    <develop>
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by Mr4k** [comment]
Okay popped open lldb and got pytorch to run in debug mode. I somehow figured out how to print MPS IR and I believe found the source of the problem:
<img width="1477" alt="Screenshot 2023-06-25 at 10 30 08 PM" src="https://github.com/pytorch/pytorch/assets/1163879/5eb43e40-bbfd-4a33-a154-c133a54c7d19">

Notice the upper bound for the random tensor of integers (si64) is 1. Looking back at the code [here](https://github.com/pytorch/pytorch/blob/58feefa4ede4779d339d76b78b0e0a3de7d47212/aten/src/ATen/native/mps/operations/Distributions.mm#L416C42-L416C42) we can see the upper bound incorrectly being set to 1.0 Quickly tested setting it to 1000000 on my machine appears to make a much better distribution. I don't think that's a very elegant fix though so I'm going to see if I can just get it to sort by float instead of int64. 

I'll try to put up a pr soon

**Comment by malfet** [comment]
Thank you very much for reporting the issue and investigating what is going on under the hood.
